sql/grant_tables/privilege_set.go

```diff
@@ -0,0 +1,614 @@
+	tblSet, ok := ps.tables[lowerTblName]
+	if !ok {
+		tblSet = PrivilegeSetTable{
+			name:    tblName,
+			privs:   make(map[PrivilegeType]struct{}),
+			columns: make(map[string]PrivilegeSetColumn),
+		}
+		ps.tables[lowerTblName] = tblSet
+	}
+	return tblSet
+}
+
+// merge merges the given set of privileges to the calling set of privileges.
+func (ps PrivilegeSetDatabase) merge(otherPs PrivilegeSetDatabase) {
+	for priv := range otherPs.privs {
+		ps.privs[priv] = struct{}{}
+	}
+	for _, otherTblSet := range otherPs.tables {
+		ps.getUseableTbl(otherTblSet.name).merge(otherTblSet)
+	}
+}
+
+// clear removes all database privileges.
+func (ps PrivilegeSetDatabase) clear() {
+	for priv := range ps.privs {
+		delete(ps.privs, priv)
+	}
+}
+
+// PrivilegeSetTable is a set containing table-level privileges.
+type PrivilegeSetTable struct {
+	name    string
+	privs   map[PrivilegeType]struct{}
+	columns map[string]PrivilegeSetColumn
+}
+
+// Name returns the name of the table that this privilege set belongs to.
+func (ps PrivilegeSetTable) Name() string {
+	return ps.name
+}
+
+// Has returns whether the given table privilege(s) exists.
+func (ps PrivilegeSetTable) Has(privileges ...PrivilegeType) bool {
+	for _, priv := range privileges {
+		if _, ok := ps.privs[priv]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// Count returns the number of table privileges.
+func (ps PrivilegeSetTable) Count() int {
+	return len(ps.privs)
+}
+
+// Column returns the set of privileges for the given column. Returns an empty set if the column does not exist.
+func (ps PrivilegeSetTable) Column(colName string) PrivilegeSetColumn {
+	colSet, ok := ps.columns[strings.ToLower(colName)]
+	if ok {
+		return colSet
+	}
+	return PrivilegeSetColumn{name: colName}
+}
+
+// GetColumns returns all columns.
+func (ps PrivilegeSetTable) GetColumns() []PrivilegeSetColumn {
+	colSets := make([]PrivilegeSetColumn, len(ps.columns))
+	i := 0
+	for _, colSet := range ps.columns {
+		colSets[i] = colSet
+		i++
+	}
+	return colSets
+}
+
+// Equals returns whether the given set of privileges is equivalent to the calling set.
+func (ps PrivilegeSetTable) Equals(otherPs PrivilegeSetTable) bool {
+	if len(ps.privs) != len(otherPs.privs) ||
+		len(ps.columns) != len(otherPs.columns) {
+		return false
+	}
+	for priv := range ps.privs {
+		if _, ok := otherPs.privs[priv]; !ok {
+			return false
+		}
+	}
+	for colName, colSet := range ps.columns {
+		if !colSet.Equals(otherPs.columns[colName]) {
+			return false
+		}
+	}
+	return true
+}
+
+// ToSlice returns all of the table privileges contained as a slice.
+func (ps PrivilegeSetTable) ToSlice() []PrivilegeType {
+	privs := make([]PrivilegeType, len(ps.privs))
+	i := 0
+	for priv := range ps.privs {
+		privs[i] = priv
+		i++
+	}
+	return privs
+}
+
+// ToSortedSlice returns all of the table privileges contained as a slice, sorted by their internal ID.
+func (ps PrivilegeSetTable) ToSortedSlice() []PrivilegeType {
```

Comment:
This should be the default behavior, and if it's too slow we should change the data structures to make it faster

Never ever rely on hash ordering

Reply:
I'm not relying on hash ordering, and there are plenty of cases where it doesn't matter. For example, if I'm converting this set of privileges into a `sql.Row`, each privilege has its own defined column, so it doesn't matter if it's sorted or not. Sure sorting doesn't waste too much time, but in cases where it truly doesn't matter, I don't see the harm in having a non-sorted option.

Reply:
I've added to the function comment stating the above.